clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/BaseIntegrationSpec.scala

```diff
@@ -80,6 +89,81 @@ abstract class BaseIntegrationSpec(clioDescription: String)
     */
   def elasticsearchUri: Uri
 
+  /** URL of vault server to use when getting bearer tokens for service accounts. */
+  private val vaultUrl = "https://clotho.broadinstitute.org:8200/"
+
+  /** Path in vault to the service account JSON to use in testing. */
+  private val vaultPath = "secret/dsde/gotc/test/clio/clio-account.json"
+
+  /** List of possible token-file locations, in order of preference. */
+  private val vaultTokenFiles = Seq(
+    new File("/etc/vault-token-dsde"),
+    new File(s"${System.getProperty("user.home")}/.vault-token")
+  )
+
+  /** Scopes needed from Google to write to our test-storage bucket. */
+  private val testStorageScopes = Seq(
+    "https://www.googleapis.com/auth/devstorage.read_write"
+  )
+
+  /**
+    * Service account credentials for use when accessing cloud resources.
+    */
+  protected lazy val serviceAccount: ServiceAccount = {
+    val vaultToken: String = vaultTokenFiles
+      .find(_.exists)
+      .map { file =>
+        new String(Files.readAllBytes(file.toPath)).stripLineEnd
+      }
+      .getOrElse {
+        sys.error("Vault token not found on filesystem!")
+      }
+
+    val vaultConfig = new VaultConfig()
+      .address(vaultUrl)
+      .token(vaultToken)
+      .build()
+
+    val vaultDriver = new Vault(vaultConfig)
+    val accountJSON =
+      vaultDriver
+        .logical()
+        .read(vaultPath)
+        .getData
+        .asScala
+        .toMap[String, String]
+        .asJson
+
+    accountJSON
+      .as[ServiceAccount]
+      .fold({ err =>
+        throw new RuntimeException(
+          s"Failed to decode service account JSON from Vault at $vaultPath",
+          err
+        )
+      }, identity)
+  }
+
```

Comment:
I think when there are only two cases, we prefer fold, especially if there's a clear "empty" case.

Reply:
Cool. Then why not just ...
```scala
accountJSON
  .as[ServiceAccount]
  .getOrElse(err => throw new RuntimeException(
    s"Failed to decode service account JSON from Vault at $vaultPath",
    err))
```

Reply:
Nevermind: That probably just returns a function on error.

Reply:
^ right, map/getOrElse wouldn't work here because getOrElse throws away the left value. 

For Options I usually reach for map/getOrElse because I built up the muscle memory for it before I learned Option.fold is a thing. I pretty much stopped using match after I started using map because I like having one pattern to think about for Futures, Options, Seqs, etc.

[Odersky weeps.](https://www.reddit.com/r/scala/comments/449td4/fold_or_pattern_matching_with_an_option/czp154s/)